src/adapter/src/catalog/apply.rs

```diff
@@ -1869,27 +1860,106 @@ fn sort_updates_inner(updates: Vec<StateUpdate>) -> Vec<StateUpdate> {
         }
     }
 
-    /// Sort item updates by [`CatalogItemId`].
+    /// Sort item updates by parsing statements to identify any [`CatalogItemId`]-based dependencies within
+    /// this set of updates and then performing a topological sort.
     fn sort_item_updates(
         item_updates: Vec<(mz_catalog::durable::Item, Timestamp, StateDiff)>,
+        state: &CatalogState,
     ) -> VecDeque<(mz_catalog::durable::Item, Timestamp, StateDiff)> {
-        item_updates
+        let mut item_id_lookup: HashMap<_, HashMap<_, _>> = HashMap::new();
+        for (item, _, _) in &item_updates {
+            item_id_lookup
+                .entry(item.schema_id)
+                .or_insert_with(HashMap::new)
+                .insert(item.name.clone(), item.global_id);
+        }
+
+        let mut items_with_dependencies = item_updates
             .into_iter()
-            // HACK: due to `ALTER SINK`, sinks can appear before the objects they
-            // depend upon. Fortunately, because sinks can never have dependencies
-            // and can never depend upon one another, to fix the topological sort,
-            // we can just always move sinks to the end.
-            .sorted_by_key(|(item, _ts, _diff)| {
-                if item.create_sql.starts_with("CREATE SINK") {
-                    CatalogItemId::User(u64::MAX)
-                } else {
-                    item.id
+            .map(|(item, ts, diff)| {
+                let parsed = mz_sql::parse::parse(&item.create_sql)
+                    .expect("failed to parse persisted SQL")
+                    .into_element()
+                    .ast;
+                let (mut id_deps, name_deps) = mz_sql::names::raw_item_dependency_ids(&parsed);
+
+                // Convert named deps into ID deps. Ideally this is empty and all dependencies are
+                // specified by ID. However, this is not the case and require some changes and a
+                // migration to fix.
+                for name in name_deps {
+                    let (db, schema, item) = match name.0.len() {
+                        3 => (
+                            Some(name.0[0].as_str()),
+                            name.0[1].as_str(),
+                            name.0[2].as_str(),
+                        ),
+                        2 => (None, name.0[0].as_str(), name.0[1].as_str()),
+                        // This must be a CTE.
+                        _ => continue,
```

Comment:
Definitely not sure, I was just experimenting to see if this would work. I think there's still CI failures, so it might not actually work. Types and functions should be fully qualified though.